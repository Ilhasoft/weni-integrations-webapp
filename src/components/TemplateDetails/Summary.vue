<template>
  <div class="whatsapp_template_summary">
    <div class="whatsapp_template_summary__title">
      <div class="whatsapp_template_summary__title__main">
        {{ $t('WhatsApp.template_details.summary.performace_overview') }}
      </div>
      <div class="whatsapp_template_summary__title__subtitle">
        {{ $t('WhatsApp.template_details.summary.last_7_days') }}
      </div>
    </div>
    <div class="whatsapp_template_summary__main">
      <div class="whatsapp_template_summary__main__card">
        <div class="whatsapp_template_summary__main__card__title">
          {{ $t('WhatsApp.template_details.summary.sent_messages') }}
          <unnnic-icon icon="info" size="sm" filled scheme="neutral-cleanest" />
        </div>
        <div class="whatsapp_template_summary__main__card__value">{{ weekValues.sent }}</div>
      </div>
      <div class="whatsapp_template_summary__main__card">
        <div class="whatsapp_template_summary__main__card__title">
          {{ $t('WhatsApp.template_details.summary.delivered_messages') }}
          <unnnic-icon icon="info" size="sm" filled scheme="neutral-cleanest" />
        </div>
        <div class="whatsapp_template_summary__main__card__value">{{ weekValues.delivered }}</div>
      </div>
      <div class="whatsapp_template_summary__main__card">
        <div class="whatsapp_template_summary__main__card__title">
          {{ $t('WhatsApp.template_details.summary.read_messages') }}
          <unnnic-icon icon="info" size="sm" filled scheme="neutral-cleanest" />
        </div>
        <div class="whatsapp_template_summary__main__card__value">
          {{ weekValues.read }} ({{
            Math.floor((weekValues.read / weekValues.sent) * 100) || '--'
          }}%)
        </div>
      </div>
    </div>
    <div class="whatsapp_template_summary__quality">
      <div class="whatsapp_template_summary__quality__title">
        <div class="whatsapp_template_summary__title__main">
          {{ $t('WhatsApp.template_details.summary.quality_overview') }}
        </div>
        <div class="whatsapp_template_summary__title__subtitle">
          {{ $t('WhatsApp.template_details.summary.last_7_days') }}
        </div>
      </div>
      <div class="whatsapp_template_summary__quality__status">
        <div class="whatsapp_template_summary__quality__status__level">
          <div class="whatsapp_template_summary__quality__status__level__title">
            {{ $t('WhatsApp.template_details.summary.current_level') }}
          </div>
          <div class="whatsapp_template_summary__quality__status__level__value">--</div>
        </div>
        <div class="whatsapp_template_summary__quality__status__blocked">
          <div class="whatsapp_template_summary__quality__status__blocked__title">
            {{ $t('WhatsApp.template_details.summary.block_reason') }}
          </div>
          <div class="whatsapp_template_summary__quality__status__blocked__value">--</div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
  import { mapState, mapActions } from 'vuex';
  export default {
    name: 'Summary',
    mounted() {
      this.fetchTemplateAnalyticsWeek();
    },
    data() {
      return {
        sevenDays: 7 * 24 * 60 * 60 * 1000,
      };
    },
    computed: {
      ...mapState({
        templateAnalytics: (state) => state.insights.templateAnalytics,
        appUuid: (state) => state.insights.appUuid,
        selectedTemplate: (state) => state.insights.selectedTemplate,
      }),
      weekValues() {
        return (
          this.templateAnalytics?.grand_totals || {
            sent: 0,
            delivered: 0,
            read: 0,
          }
        );
      },
    },
    methods: {
      ...mapActions('insights', ['getTemplateAnalytics', 'getTemplates']),
      fetchTemplateAnalyticsWeek() {
        const params = {
          app_uuid: this.appUuid,
          filters: {
            start: this.formatDate(new Date(Date.now() - this.sevenDays)),
            end: this.formatDate(new Date()),
            fba_template_ids: this.selectedTemplate.translations.map(
              (item) => item.message_template_id,
            ),
          },
        };
        this.getTemplateAnalytics(params);
      },
      formatDate(date) {
        return `${date.getMonth() + 1}-${date.getDate()}-${date.getFullYear()}`;
      },
    },
  };
</script>

<style lang="scss">
  @import './styles.scss';
</style>
