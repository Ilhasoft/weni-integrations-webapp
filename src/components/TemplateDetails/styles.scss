.whatsapp_template_summary {
  &__title {
    display: flex;
    flex-direction: row;
    gap: $unnnic-spacing-sm;
    align-items: center;
    font-family: $unnnic-font-family-secondary;
    &__main {
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-bold;
      color: $unnnic-color-neutral-darkest;
    }
    &__subtitle {
      font-size: $unnnic-font-size-body-md;
      color: $unnnic-color-neutral-cloudy;
    }
  }

  &__main {
    display: flex;
    flex-direction: row;
    gap: $unnnic-spacing-sm;
    margin-top: $unnnic-spacing-xs;

    &__card {
      display: flex;
      flex-direction: column;
      gap: $unnnic-spacing-nano;
      font-family: $unnnic-font-family-secondary;
      color: $unnnic-color-neutral-darkest;

      &__title {
        display: flex;
        font-size: $unnnic-font-size-body-gt;
        align-items: center;
      }

      &__value {
        font-size: $unnnic-font-size-title-lg;
      }
    }
  }

  &__quality {
    margin-top: $unnnic-spacing-xs;
    &__title {
      display: flex;
      flex-direction: row;
      gap: $unnnic-spacing-sm;
      align-items: center;
    }
    &__status {
      display: flex;
      flex-direction: row;
      gap: $unnnic-spacing-sm;
      align-items: center;
      margin-top: $unnnic-spacing-xs;
      
      &__level {
        display: flex;
        flex-direction: column;
        font-family: $unnnic-font-family-secondary;
        color: $unnnic-color-neutral-darkest;
        &__title {
          font-size: $unnnic-font-size-body-gt;
          align-items: center;
        }
  
        &__value {
          font-size: $unnnic-font-size-title-lg;
        }

      }

      &__blocked{
        display: flex;
        flex-direction: column;
        font-family: $unnnic-font-family-secondary;
        color: $unnnic-color-neutral-darkest;
        &__title {
          font-size: $unnnic-font-size-body-gt;
        }
  
        &__value {
          font-size: $unnnic-font-size-title-lg;
        }
      }
    }
  }
}
