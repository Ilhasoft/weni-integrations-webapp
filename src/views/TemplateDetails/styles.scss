.wpp_template_details{
    display: flex;
    flex-direction: column;
    width: 100%;
    gap: 1rem;

    &__header{
        display: flex;
        flex-direction: row;
        align-items: center;
    }
}