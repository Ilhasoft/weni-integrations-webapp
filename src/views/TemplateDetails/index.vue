<template>
  <div class="wpp_template_details">
    <!-- Breadcrumb -->
    <div class="wpp_template_details__breadcrumb">
      <unnnic-breadcrumb
        :crumbs="[
          {
            name: $t('WhatsApp.template_details.crumbs.my_apps'),
            path: '/apps/my',
            meta: $t('WhatsApp.template_details.crumbs.my_apps'),
          },
          {
            name: $t('WhatsApp.template_details.crumbs.manage_templates'),
            path: `/apps/my/wpp-cloud/${app_uuid}/templates`,
            meta: $t('WhatsApp.template_details.crumbs.manage_templates'),
          },
          {
            name: $t('WhatsApp.template_details.crumbs.model_details'),
            path: '/',
            meta: $t('WhatsApp.template_details.crumbs.model_details'),
          },
        ]"
        @crumbClick="redirectTo"
      />
    </div>
    <!-- Header -->
    <div class="wpp_template_details__header">
      <div class="wpp_template_details__header__title">
        <div class="wpp_template_details__header__title__text">{{ selectedTemplate.name }}</div>
        <div class="wpp_template_details__header__about__title__tag">
          <unnnic-tag scheme="neutral-dark" text="Marketing" />
        </div>
      </div>
      <div class="wpp_insights__header__button">
        <unnnic-button
          text="Editar Modelo"
          ref="wpp_template_details__button__edit"
          type="secondary"
        />
      </div>
    </div>
    <!-- Tabs -->
    <div class="wpp_insights__tabs">
      <unnnic-tab initialTab="first" :tabs="['first', 'second']">
        <template slot="tab-head-first">{{ $t('WhatsApp.template_details.summary') }}</template>
        <template slot="tab-panel-first">
          <Summary />
        </template>
        <template slot="tab-head-second">Insights</template>
        <template slot="tab-panel-second">
          <Insights />
        </template>
      </unnnic-tab>
    </div>
  </div>
</template>

<script>
  import Insights from '../Insights/index.vue';
  import Summary from '../../components/TemplateDetails/Summary.vue';
  import { mapState } from 'vuex';
  export default {
    name: 'AppDetails',
    components: {
      Insights,
      Summary,
    },
    data() {
      return {
        route: this.$route,
        crumb_title: 'Detalhes do modelo',
      };
    },
    computed: {
      ...mapState({
        selectedTemplate: (state) => state.insights.selectedTemplate,
        app_uuid: (state) => state.insights.appUuid,
      }),
    },
    methods: {
      redirectTo(crumb) {
        if (
          crumb.meta === this.$t('WhatsApp.template_details.crumbs.model_details') ||
          crumb.meta === ''
        ) {
          return;
        }
        this.$router.push(crumb.path);
      },
    },
  };
</script>
<style lang="scss">
  @import './styles.scss';
</style>
