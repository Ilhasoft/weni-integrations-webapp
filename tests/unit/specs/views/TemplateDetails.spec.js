import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import TemplateDetails from '@/views/TemplateDetails/index.vue';
import { singleApp } from '../../../__mocks__/appMock';

const genericApp = {
  ...singleApp,
  name: 'generic',
  code: 'generic',
  generic: true,
  config: { channel_name: 'A random generic' },
};

const localVue = createLocalVue();
localVue.use(Vuex);

describe('TemplateDetails/index.vue', () => {
  let wrapper;
  let actions;
  let state;
  let store;
  const crumb = {
    name: 'my apps',
    path: '/apps/my',
    meta: 'my apps',
  };

  beforeEach(() => {
    actions = {
      getTemplates: jest.fn(),
      getTemplateAnalytics: jest.fn(),
    };

    state = {
      myApps: {
        configuredApps: [singleApp, genericApp],
        loadingConfiguredApps: false,
        errorConfiguredApps: null,
        installedApps: [singleApp, genericApp],
        loadingInstalledApps: false,
        errorInstalledApps: null,
      },
      insights: {
        isActive: false,
        templateAnalytics: {
          data: [
            {
              template_id: '730081812069736',
              template_name: null,
              totals: { sent: 51, delivered: 51, read: 30 },
              dates: [
                { start: '2024-01-04', sent: 11, delivered: 10, read: 6 },
                { start: '2024-01-05', sent: 7, delivered: 7, read: 5 },
                { start: '2024-01-06', sent: 6, delivered: 5, read: 1 },
                { start: '2024-01-07', sent: 4, delivered: 5, read: 4 },
                { start: '2024-01-08', sent: 1, delivered: 2, read: 2 },
                { start: '2024-01-09', sent: 0, delivered: 0, read: 0 },
                { start: '2024-01-10', sent: 11, delivered: 11, read: 5 },
                { start: '2024-01-11', sent: 11, delivered: 11, read: 7 },
              ],
            },
          ],
          grand_totals: { sent: 51, delivered: 51, read: 30 },
        },
        loadingTemplateAnalytics: false,
        errorTemplateAnalytics: null,
        selectedTemplate: {
          uuid: 'd9019398-e5df-421f-98d5-8cfa711e7b20',
          name: 'testeana2',
          created_on: '2024-01-10 17:10:01.952251+00:00',
          category: 'MARKETING',
          translations: [],
        },
        appUuid: null,
        templates: {
          count: 1,
          next: null,
          previous: null,
          results: [
            {
              uuid: 'd9019398-e5df-421f-98d5-8cfa711e7b20',
              name: 'testeana2',
              created_on: '2024-01-10 17:10:01.952251+00:00',
              category: 'MARKETING',
              translations: [],
            },
          ],
        },
        errorTemplates: null,
      },
    };

    store = new Vuex.Store({
      actions,
      state,
    });

    wrapper = shallowMount(TemplateDetails, {
      localVue,
      store,
      mocks: {
        $t: () => 'some specific text',
        $router: {
          push: jest.fn(),
        },
        $route: {
          path: '/template-details',
        },
      },
    });
  });

  it('should be rendered properly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should call redirectTo()', () => {
    const spy = spyOn(wrapper.vm, 'redirectTo');
    expect(spy).not.toHaveBeenCalled();
    wrapper.vm.redirectTo(crumb);
    expect(spy).toHaveBeenCalledTimes(1);
  });
  describe('redirectTo', () => {
    it('should change route to my apps', () => {
      const spy = spyOn(wrapper.vm.$router, 'push');
      wrapper.vm.redirectTo(crumb);

      expect(spy).toBeCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(`/apps/my`);
    });
    it('should not change route', () => {
      const thisCrumb = {
        name: '',
        path: '/',
        meta: '',
      };
      const spy = spyOn(wrapper.vm.$router, 'push');
      wrapper.vm.redirectTo(thisCrumb);

      expect(spy).not.toBeCalled();
    });
  });
  describe('redirectEdit', () => {
    it('should change route to edit template', () => {
      const spy = spyOn(wrapper.vm.$router, 'push');
      expect(spy).not.toBeCalled();
      wrapper.vm.redirectEdit();
      expect(spy).toBeCalledTimes(1);
    });
  });
});
