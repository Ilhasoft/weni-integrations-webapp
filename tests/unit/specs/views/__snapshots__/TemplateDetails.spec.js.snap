// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateDetails/index.vue should be rendered properly 1`] = `
<div class="wpp_template_details">
  <div class="wpp_template_details__breadcrumb">
    <unnnic-breadcrumb-stub crumbs="[object Object],[object Object],[object Object]"></unnnic-breadcrumb-stub>
  </div>
  <div class="wpp_template_details__header">
    <div class="wpp_template_details__header__title">
      <div class="wpp_template_details__header__title__text">testeana2</div>
      <div class="wpp_template_details__header__about__title__tag">
        <unnnic-tag-stub type="default" text="Marketing" count="0" scheme="neutral-dark"></unnnic-tag-stub>
      </div>
    </div>
    <div class="wpp_insights__header__button">
      <unnnic-button-stub size="large" text="Editar Modelo" type="secondary"></unnnic-button-stub>
    </div>
  </div>
  <div class="wpp_insights__tabs">
    <unnnic-tab-stub size="md" initialtab="first" tabs="first,second"><template></template> <template></template> <template></template> <template></template></unnnic-tab-stub>
  </div>
</div>
`;
